server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java

```diff
@@ -1252,39 +1254,60 @@ protected void runInContext() {
-                                            RouterHealthChecksFailuresToRestartVrCK + ", attempting restart of router.");
-                                    recreateRouter(router.getId());
-                                }
-                            }
-                        }
-                    }
+                    List<String> failingChecks = getFailingChecks(router, answer);
+                    handleFailingChecks(router, failingChecks);
                 }
             } catch (final Exception ex) {
                 s_logger.error("Fail to complete the FetchRouterHealthChecksResultTask! ", ex);
                 ex.printStackTrace();
             }
         }
+
+        private List<String> getFailingChecks(DomainRouterVO router, GetRouterMonitorResultsAnswer answer) {
+            if (answer == null) {
+                s_logger.warn("Unable to fetch monitor results for router " + router);
+                updateRouterConnectivityHealthCheck(router.getId(), false, "Communication failed");
+                return Arrays.asList(CONNECTIVITY_TEST);
+            } else if (!answer.getResult()) {
+                s_logger.warn("Failed to fetch monitor results from router " + router + " with details: " + answer.getDetails());
+                updateRouterConnectivityHealthCheck(router.getId(), false, "Failed to fetch results with details: " + answer.getDetails());
+                return Arrays.asList(CONNECTIVITY_TEST);
+            } else {
+                updateRouterConnectivityHealthCheck(router.getId(), true, "Successfully fetched data");
+                updateDbHealthChecksFromRouterResponse(router.getId(), answer.getMonitoringResults());
+                return answer.getFailingChecks();
+            }
+        }
+
+        private void handleFailingChecks(DomainRouterVO router, List<String> failingChecks) {
+            if (failingChecks == null || failingChecks.size() == 0) {
+                return;
+            }
+
+            String alertMessage = "Health checks failed: " + failingChecks.size() + " failing checks on router " + router.getUuid();
+            _alertMgr.sendAlert(AlertType.ALERT_TYPE_DOMAIN_ROUTER, router.getDataCenterId(), router.getPodIdToDeployIn(),
+                    alertMessage, alertMessage);
+            s_logger.warn(alertMessage + ". Checking failed health checks to see if router needs reboot");
+
+            String checkFailsToRecreateVr = RouterHealthChecksFailuresToRecreateVr.valueIn(router.getDataCenterId());
+            StringBuilder failingChecksEvent = new StringBuilder("Router " + router.getUuid() + " has failing checks:");
+            boolean recreateRouter = false;
+            for (String failedCheck : failingChecks) {
+                failingChecksEvent.append(" ").append(failedCheck);
+                if (StringUtils.isNotBlank(checkFailsToRecreateVr) && checkFailsToRecreateVr.contains(failedCheck)) {
+                    recreateRouter = true;
+                }
+            }
```

Comment:
check if a recreate worthy failure has happened

Reply:
Don't understand the question @DaanHoogland . That's decided by the global config that I am reading in here to set the flag recreateRouter.